bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.automation.jsscripting.internal;
+
+import java.util.Objects;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.automation.module.script.action.ScriptExecution;
+import org.openhab.core.scheduler.Scheduler;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * Utility class for providing easy access to script services.
+ *
+ * @author Florian Hotze - Initial contribution
+ */
+@Component(immediate = true)
+@NonNullByDefault
+public class JSScriptServiceUtil {
+    private static @Nullable Scheduler scheduler;
```

Comment:
This is a slightly odd use of OSGi. You have a component which is activated but then permanently stores (as static) the passed in references. Typically these references would not be static, however it means that consumers of this class would need it injecting too, rather that using static references to it.

I'm not sure it (a) this is a common pattern in openHAB (in which case go ahead with it), or (b) if it will cause any real problems or not. There _may_ be a problem with plugin reloading here, as the component will be reloaded and the fields will then be reset to new values, and any consumers (via the static ref) will see updated values, whereas consumers via injection will see two versions of `JSScriptServiceUtil` where the old one internally gets updated to the values of the new one. I am also not sure if there could be issues with components hanging on to static class references (of `JSScriptServiceUtil`) and hence preventing class collection by the GC.

The alternative to all this would be to (a) make these references non-static, (b) make `ThreadsafeTimers` a component and inject this class as a `@Reference` into it, and (c) fetch an instance of `ThreadsafeTimers` from the bundle context (rather than creating your own).

Personally I think this is a better way to go, however if this is an existing pattern in openHAB then I don't have an issue with it.

Reply:
Oh, and doing without static refs means that you can get rid of the `@Nullable` annotations as the references can never be null.

Reply:
I agree with @jpg0. This is an ugly workaround that is used in the DSL Script engine because it would (at least) require very heavy refactoring to make the OSGi components available to XText. It should not be used if there is a better way.

Reply:
Could I also make `JSScriptServiceUtil` non-static, initialize it from somewhere (OpenHABGraalJSScriptEngine) and use the constructor to get `Scheduler` and `ScriptExecution` from OSGi?

Reply:
@jpg0 @J-N-K See commit https://github.com/openhab/openhab-addons/pull/13695/commits/fedbbe61fbbfec6b90c0e58692f1d40b06d31dbd for how I am retrieving the OSGi components.

Reply:
Your updated method is slightly better, but why not just inject the dependencies? The reason for injection is to break the coupling between classes, to allow things like testing them independently. Putting all the lookup code in the constructor now ties the class to OSGi too, as well as making it more complex. Are you doing that to make the lookup from JS easier? (There is an OSGi lookup ability in JS already.)

Reply:
> why not just inject the dependencies?

You are referring to the solution you propose above, correctly?

> The alternative to all this would be to (a) make these references non-static, (b) make ThreadsafeTimers a component and inject this class as a @Reference into it, and (c) fetch an instance of ThreadsafeTimers from the bundle context (rather than creating your own).

So my question is, how to get the lock object into ThreadsafeTimers when I fetch it from the bundle context.

Reply:
You could inject the `ScriptExecution` to the `GraalJSScriptEngineFactory` which is already a component and then pass it in the constructor to the `OpenhabGraalJSScriptEngine`. 

A "better" but more complex but better solution would be to make a factory component that creates a thread-safe component instance for each caller. The factory component could then get the `ScriptExecution` and also the `Scheduler` via injection and create a new lock object each time it creates a new instance.

Reply:
> You could inject the ScriptExecution to the GraalJSScriptEngineFactory which is already a component and then pass it in the constructor to the OpenhabGraalJSScriptEngine.

Can I also inject the `JSScriptServiceUtil` (as long as it is a `@Component`) to the `GraalJSScriptEngineFactory`?

> A "better" but more complex but better solution would be to make a factory component that creates a thread-safe component instance for each caller. The factory component could then get the ScriptExecution and also the Scheduler via injection and create a new lock object each time it creates a new instance.

I am not sure if I able to implement this. I will try.

Reply:
If you need assistance, feel free to contact me by private mail.

Reply:
> If you need assistance, feel free to contact me by private mail.

Thanks for the offer.

Could you please check if my current implementation is okay?

Regarding the "better" solution: I am asking myself if is worth the additional work.
If we need access to more OSGi service, we can just extend `JSScriptServiceUtil` and do not have to change the injection.

Reply:
Much better. I have made some minor suggestions in a PR to your fork.

Reply:
@J-N-K See https://github.com/openhab/openhab-addons/pull/13695/commits/77bc18a8da5cd5615837ec561b0f3b5500e7f663 for the changes I spoke about in your PR.

Reply:
LGTM

Reply:
Okay, could you then please give an approving review?